src/methods/properties.js

```diff
@@ -102,48 +102,64 @@ function InternalSetProperty(realm: Realm, O: ObjectValue, P: PropertyKeyValue,
   propertyBinding.descriptor = desc;
 }
 
-function InternalUpdatedProperty(realm: Realm, O: ObjectValue, P: PropertyKeyValue, oldDesc?: Descriptor) {
+function InternalEmitPropertyMutationEffect(
+  realm: Realm,
+  Target: null | ObjectValue | AbstractObjectValue,
+  P: PropertyKeyValue,
+  desc: void | Descriptor,
+  oldDesc: void | Descriptor
+) {
+  if (Target === null) return;
   let generator = realm.generator;
   if (!generator) return;
-  if (!O.isIntrinsic() && O.temporalAlias === undefined) return;
+  if (Target instanceof AbstractObjectValue) {
+    // If the target is conditional, we pass the original object as target.
+    // If it is coming from some other operation, it should have some intrinsic
+    // name. In the future, we may have other ways to create abstract objects
+    // and at that point we can relax this invariant to include those cases too.
+    invariant(Target.isIntrinsic());
+  } else {
+    invariant(Target instanceof ObjectValue);
+    if (!Target.isIntrinsic() && Target.temporalAlias === undefined) {
+      // We only emit effects on intrinsics or snapshots.
+      return;
+    }
+    invariant(!Target.mightBeLeakedObject()); // leaked objects are never updated
+    invariant(!Target.mightBeFinalObject()); // final objects are never updated
+  }
   if (P instanceof SymbolValue) return;
   if (P instanceof StringValue) P = P.value;
-  invariant(!O.mightBeLeakedObject()); // leaked objects are never updated
-  invariant(!O.mightBeFinalObject()); // final objects are never updated
   invariant(typeof P === "string");
-  let propertyBinding = InternalGetPropertiesMap(O, P).get(P);
-  invariant(propertyBinding !== undefined); // The callers ensure this
-  let desc = propertyBinding.descriptor;
```

Comment:
The reason I don't read it from the current binding is so that I can in the future represent weak updates on the binding without actually emitting a weak descriptor. So instead I explicitly pass it in. That also lets me only pass in the target rather than the modified object.